src/solvers/dgsem_tree/indicators_2d.jl

```diff
@@ -190,6 +190,290 @@ function (löhner::IndicatorLöhner)(u::AbstractArray{<:Any,4},
+      var_neighbor_left  = indicator_IDP.variable(u_local_left,  equations)
+      var_neighbor_right = indicator_IDP.variable(u_local_right, equations)
+
+      var_min[index_right...] = min(var_min[index_right...], var_neighbor_left)
+      var_max[index_right...] = max(var_max[index_right...], var_neighbor_left)
+
+      var_min[index_left...] = min(var_min[index_left...], var_neighbor_right)
+      var_max[index_left...] = max(var_max[index_left...], var_neighbor_right)
+    end
+  end
+
+  # Loop over L2 mortars
+  for mortar in eachmortar(dg, cache)
+    # Get neighboring element ids
+    lower = cache.mortars.neighbor_ids[1, mortar]
+    upper = cache.mortars.neighbor_ids[2, mortar]
+    large = cache.mortars.neighbor_ids[3, mortar]
+
+    # buffer variables. TODO: Spare these unnecessary allocations
+    u_tmp_upper  = similar(view(u_old, :, 1, :, large))
+    u_tmp_lower  = similar(u_tmp_upper)
+    u_tmp_large1 = similar(u_tmp_upper)
+    u_tmp_large2 = similar(u_tmp_upper)
+
+    # Interpolate element face data to adjacent interface locations and use for var_max/min
+    if cache.mortars.large_sides[mortar] == 1 # -> large element on left side
+      if cache.mortars.orientations[mortar] == 1
+        # L2 mortars in x-direction
+        u_large = view(u_old, :, nnodes(dg), :, large)
+        u_upper = view(u_old, :,          1, :, upper)
+        u_lower = view(u_old, :,          1, :, lower)
+
+        multiply_dimensionwise!(u_tmp_upper, dg.mortar.forward_upper, u_large)
+        multiply_dimensionwise!(u_tmp_lower, dg.mortar.forward_lower, u_large)
+
+        multiply_dimensionwise!(u_tmp_large1, dg.mortar.reverse_upper, u_upper)
+        multiply_dimensionwise!(u_tmp_large2, dg.mortar.reverse_lower, u_lower)
+
+        for i in eachnode(dg)
+          # large to small
+          var_min[1, i, upper] = min(var_min[1, i, upper], indicator_IDP.variable(view(u_tmp_upper, :, i), equations))
+          var_max[1, i, upper] = max(var_max[1, i, upper], indicator_IDP.variable(view(u_tmp_upper, :, i), equations))
+
+          var_min[1, i, lower] = min(var_min[1, i, lower], indicator_IDP.variable(view(u_tmp_lower, :, i), equations))
+          var_max[1, i, lower] = max(var_max[1, i, lower], indicator_IDP.variable(view(u_tmp_lower, :, i), equations))
+
+          # small to large
+          if i <= nnodes(dg)/2
+            var_min[nnodes(dg), i, large] = min(var_min[nnodes(dg), i, large], indicator_IDP.variable(view(u_tmp_large2, :, i), equations))
+            var_max[nnodes(dg), i, large] = max(var_max[nnodes(dg), i, large], indicator_IDP.variable(view(u_tmp_large2, :, i), equations))
+          else
+            var_min[nnodes(dg), i, large] = min(var_min[nnodes(dg), i, large], indicator_IDP.variable(view(u_tmp_large1, :, i), equations))
+            var_max[nnodes(dg), i, large] = max(var_max[nnodes(dg), i, large], indicator_IDP.variable(view(u_tmp_large1, :, i), equations))
+          end
```

Comment:
It looks like large parts of the code are the same, aren't they? If so, you could move them into a function and just keep the logic setting `u_large`, `u_upper`, and `u_lower` separate.

Reply:
Have a look at `element_solutions_to_mortars!` and `mortar_fluxes_to_elements!` for a similar approach.

Reply:
Yes, thanks. I implemented a function which includes the four cases. But since I need the indices `(nnodes(dg), i)` and `(nnodes(dg), :)` in all combinations, I now have functions `index_large = i -> (nnodes(dg), i)` as parameters for this function, which produces many allocations. Do you have an alternative idea?

Rn, my code is
```
# Interpolate element face data to adjacent interface locations and use for var_max/min
    if cache.mortars.large_sides[mortar] == 1 # -> large element on left side
      if cache.mortars.orientations[mortar] == 1
        # L2 mortars in x-direction
        index_large = i -> (nnodes(dg), i)
        index_small = i -> (1, i)
      else
        # L2 mortars in y-direction
        index_large = i -> (i, nnodes(dg))
        index_small = i -> (i, 1)
      end
    else # large_sides[mortar] == 2 -> large element on right side
      if cache.mortars.orientations[mortar] == 1
        # L2 mortars in x-direction
        index_large = i -> (1, i)
        index_small = i -> (nnodes(dg), i)
      else
        # L2 mortars in y-direction
        index_large = i -> (i, 1)
        index_small = i -> (i, nnodes(dg))
      end
    end

    element_solutions_to_mortars!(u_old, indicator_IDP, dg, equations,
                                  large, upper, lower, index_large, index_small)
```
with
```
@inline function element_solutions_to_mortars!(u_old::AbstractArray{<:Any,4}, indicator_IDP, dg, equations,
                                               large, upper, lower,
                                               index_large, index_small)

  @unpack var_max, var_min = indicator_IDP.cache.ContainerShockCapturingIndicator
  # buffer variables. TODO: Spare these unnecessary allocations
  u_tmp_upper  = similar(view(u_old, :, 1, :, large))
  u_tmp_lower  = similar(u_tmp_upper)
  u_tmp_large1 = similar(u_tmp_upper)
  u_tmp_large2 = similar(u_tmp_upper)

  u_large = view(u_old, :, index_large(:)..., large)
  u_upper = view(u_old, :, index_small(:)..., upper)
  u_lower = view(u_old, :, index_small(:)..., lower)

  multiply_dimensionwise!(u_tmp_upper, dg.mortar.forward_upper, u_large)
  multiply_dimensionwise!(u_tmp_lower, dg.mortar.forward_lower, u_large)

  multiply_dimensionwise!(u_tmp_large1, dg.mortar.reverse_upper, u_upper)
  multiply_dimensionwise!(u_tmp_large2, dg.mortar.reverse_lower, u_lower)

  for i in eachnode(dg)
    # large to small
    var_min[index_small(i)..., upper] = min(var_min[index_small(i)..., upper], indicator_IDP.variable(view(u_tmp_upper, :, i), equations))
    var_max[index_small(i)..., upper] = max(var_max[index_small(i)..., upper], indicator_IDP.variable(view(u_tmp_upper, :, i), equations))

    var_min[index_small(i)..., lower] = min(var_min[index_small(i)..., lower], indicator_IDP.variable(view(u_tmp_lower, :, i), equations))
    var_max[index_small(i)..., lower] = max(var_max[index_small(i)..., lower], indicator_IDP.variable(view(u_tmp_lower, :, i), equations))

    # small to large
    if i <= nnodes(dg)/2
      var_min[index_large(i)..., large] = min(var_min[index_large(i)..., large], indicator_IDP.variable(view(u_tmp_large2, :, i), equations))
      var_max[index_large(i)..., large] = max(var_max[index_large(i)..., large], indicator_IDP.variable(view(u_tmp_large2, :, i), equations))
    else
      var_min[index_large(i)..., large] = min(var_min[index_large(i)..., large], indicator_IDP.variable(view(u_tmp_large1, :, i), equations))
      var_max[index_large(i)..., large] = max(var_max[index_large(i)..., large], indicator_IDP.variable(view(u_tmp_large1, :, i), equations))
    end
  end

  return nothing
end
```

Reply:
Because I need `index_large(i)` below. Or do I miss something?

Reply:
Either extract a `view` indexed by `index_large = (nnodes(dg), :)` or use broadcasting? Could you mock up a minimal example using the technique where it's easier to see?